core/src/memory.rs

```
@@ -367,49 +543,80 @@ impl AllocationsContext {
-                if (*page).saturating_sub(previous) > pages {
-                    at = Some((previous).saturating_add(1.into()));
-                    break;
-                }
-            }
+            previous = page.inc().map_err(|_| Error::OutOfBounds)?;
         }
 
-        let at = at
-            .or_else(|| current.map(|v| (*v).saturating_add(1.into())))
-            .unwrap_or(self.static_pages);
+        let (start, not_grown) = if let Some(start) = start {
+            (start, pages)
+        } else {
+            // If we cannot find interval between already allocated pages, then try to alloc new pages.
+
+            // Panic is safe, because we check, that last allocated page can be incremented in loop above.
+            let start = self
+                .allocations
+                .last()
+                .map(|last| last.inc().unwrap_or_else(|err| {
+                    unreachable!("Cannot increment last allocation: {}, but we checked in loop above that it can be done", err)
+                }))
+                .unwrap_or(self.static_pages);
+            let end = start.add(pages).map_err(|_| Error::OutOfBounds)?;
+            if end > self.max_pages {
+                return Err(Error::OutOfBounds);
+            }
 
-        let final_page = at.saturating_add(pages);
-        if final_page > self.max_pages {
-            return Err(Error::OutOfBounds);
-        }
+            // Panic is safe, because in loop above we checked it.
+            let extra_grow = end.sub(mem_size).unwrap_or_else(|err| {
+                unreachable!(
+                    "`mem_size` must be bigger than all allocations or static pages, but get {}",
+                    err
+                )
+            });
+
+            // Panic is safe, in other case we would found interval inside existing memory.
+            if extra_grow == WasmPageNumber::zero() {
+                unreachable!("`extra grow cannot be zero");
```

why should we panic here instead of returning error? seems we can somehow run into this branch calling alloc(0) ?

---

No we can't. In case alloc(0)  we will find an answer in loop

---

Do we have a test for that?

---

In benchmarks only currently, will make one unit also

---

Done